rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeper.java

```diff
@@ -296,6 +297,43 @@ public void onRemoval(
                 .build();
     }
 
+    @Override
+   public boolean readData(String path, long offset, int length, long fileSize, long lastModified, int clusterType)
+           throws TException
+   {
+       int blockSize = CacheConfig.getBlockSize(conf);
+       byte[] buffer = new byte[blockSize];
+       BookKeeperFactory bookKeeperFactory = new BookKeeperFactory(this);
+       CachingNativeS3FileSystem fs = null;
+       FSDataInputStream inputStream = null;
+       long startBlock = offset / blockSize;
+       long endBlock = ((offset + (length - 1)) / CacheConfig.getBlockSize(conf)) + 1;
+       try {
+           int idx = 0;
+           List<BlockLocation> blockLocations = getCacheStatus(path, fileSize, lastModified, startBlock, endBlock, clusterType);
+
+           for (int blockNum = (int) startBlock; blockNum < endBlock; blockNum++, idx++) {
+               int readStart = blockNum * blockSize;
+               if (blockLocations.get(idx).getLocation() == Location.LOCAL) {
```

Comment:
change `blockLocations.get(idx).getLocation() == Location.LOCAL` to `blockLocations.get(idx).getLocation() != Location.CACHED`